File: clients/csharp/SchmiTech.Orbit.Client/ApiClient.cs
using System.Buffers;
using System.Net.Http.Headers;
using System.Runtime.CompilerServices;
using System.Text;
using System.Text.Json;

namespace SchmiTech.Orbit;

public record StreamResponse(string Text, bool Done);

public class ApiClient
{
    private readonly string _apiUrl;
    private readonly string? _apiKey;
    private string? _sessionId;
    private readonly HttpClient _http;

    public ApiClient(string apiUrl, string? apiKey = null, string? sessionId = null)
    {
        _apiUrl = apiUrl ?? throw new ArgumentNullException(nameof(apiUrl));
        _apiKey = apiKey;
        _sessionId = sessionId;
        _http = new HttpClient(new SocketsHttpHandler { PooledConnectionLifetime = TimeSpan.FromMinutes(5) })
        {
            Timeout = Timeout.InfiniteTimeSpan
        };
    }

    private string Endpoint => _apiUrl.EndsWith("/v1/chat") ? _apiUrl : _apiUrl.TrimEnd('/') + "/v1/chat";

    public void SetSessionId(string? sessionId) => _sessionId = sessionId;
    public string? GetSessionId() => _sessionId;

    private HttpRequestMessage BuildRequest(string message, bool stream)
    {
        var req = new HttpRequestMessage(HttpMethod.Post, Endpoint);
        req.Headers.Accept.Clear();
        req.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue(stream ? "text/event-stream" : "application/json"));
        if (!string.IsNullOrEmpty(_apiKey)) req.Headers.Add("X-API-Key", _apiKey);
        if (!string.IsNullOrEmpty(_sessionId)) req.Headers.Add("X-Session-ID", _sessionId);
        req.Headers.Add("X-Request-ID", DateTimeOffset.UtcNow.ToUnixTimeMilliseconds().ToString());

        var body = new
        {
            messages = new[] { new { role = "user", content = message } },
            stream
        };
        req.Content = new StringContent(JsonSerializer.Serialize(body), Encoding.UTF8, "application/json");
        return req;
    }

    public async IAsyncEnumerable<StreamResponse> StreamChatAsync(string message, bool stream, [EnumeratorCancellation] CancellationToken ct = default)
    {
        using var req = BuildRequest(message, stream);
        using var resp = await _http.SendAsync(req, HttpCompletionOption.ResponseHeadersRead, ct);
        resp.EnsureSuccessStatusCode();

        if (!stream)
        {
            var json = await resp.Content.ReadAsStringAsync(ct);
            using var doc = JsonDocument.Parse(json);
            if (doc.RootElement.TryGetProperty("response", out var r))
            {
                yield return new StreamResponse(r.GetString() ?? string.Empty, true);
                yield break;
            }
            yield return new StreamResponse(json, true);
            yield break;
        }

        await using var s = await resp.Content.ReadAsStreamAsync(ct);
        var reader = new StreamReader(s, Encoding.UTF8);
        string? line;
        while ((line = await reader.ReadLineAsync()) != null)
        {
            ct.ThrowIfCancellationRequested();
            line = line.Trim();
            if (line.Length == 0) continue;
            if (line.StartsWith("data: "))
            {
                var payload = line[6..].Trim();
                if (payload.Length == 0 || payload == "[DONE]")
                {
                    yield return new StreamResponse(string.Empty, true);
                    yield break;
                }
                try
                {
                    using var doc = JsonDocument.Parse(payload);
                    if (doc.RootElement.TryGetProperty("error", out var err) && err.TryGetProperty("message", out var msg))
                        throw new Exception(msg.GetString());
                    bool done = doc.RootElement.TryGetProperty("done", out var d) && d.GetBoolean();
                    if (doc.RootElement.TryGetProperty("response", out var t))
                        yield return new StreamResponse(t.GetString() ?? string.Empty, done);
                    if (done) { yield return new StreamResponse(string.Empty, true); yield break; }
                }
                catch
                {
                    yield return new StreamResponse(payload, false);
                }
            }
            else
            {
                yield return new StreamResponse(line, false);
            }
        }
    }
}

File: clients/kotlin/bin/main/com/schmitech/orbit/ApiClient.kt
package com.schmitech.orbit

import okhttp3.*
import okio.BufferedSource
import java.io.IOException

data class StreamResponse(val text: String, val done: Boolean)

class ApiClient(private val apiUrl: String, private val apiKey: String? = null, private var sessionId: String? = null) {
    private val client = OkHttpClient.Builder().build()

    private fun endpoint(): String = if (apiUrl.endsWith("/v1/chat")) apiUrl else apiUrl.trimEnd('/') + "/v1/chat"

    fun setSessionId(id: String?) { sessionId = id }

    private fun requestBody(message: String, stream: Boolean): RequestBody {
        val json = "{" +
                "\"messages\":[{\"role\":\"user\",\"content\":${escapeJson(message)}}]," +
                "\"stream\":$stream}"
        return json.toRequestBody("application/json".toMediaType())
    }

    private fun escapeJson(s: String): String {
        val sb = StringBuilder("\"")
        for (c in s) {
            when (c) {
                '\\' -> sb.append("\\\\")
                '"' -> sb.append("\\\"")
                '\n' -> sb.append("\\n")
                '\r' -> sb.append("\\r")
                '\t' -> sb.append("\\t")
                else -> sb.append(c)
            }
        }
        sb.append('"')
        return sb.toString()
    }

    fun streamChat(message: String, stream: Boolean = true, onChunk: (StreamResponse) -> Unit) {
        val reqBuilder = Request.Builder()
            .url(endpoint())
            .post(requestBody(message, stream))
            .addHeader("Content-Type", "application/json")
            .addHeader("Accept", if (stream) "text/event-stream" else "application/json")
            .addHeader("X-Request-ID", System.currentTimeMillis().toString())
        if (!apiKey.isNullOrEmpty()) reqBuilder.addHeader("X-API-Key", apiKey)
        if (!sessionId.isNullOrEmpty()) reqBuilder.addHeader("X-Session-ID", sessionId!!)

        client.newCall(reqBuilder.build()).enqueue(object: Callback {
            override fun onFailure(call: Call, e: IOException) {
                throw e
            }

            override fun onResponse(call: Call, response: Response) {
                response.use {
                    if (!response.isSuccessful) throw IOException("HTTP ${'$'}{response.code}")
                    val src: BufferedSource = response.body!!.source()
                    if (!stream) {
                        val text = response.body!!.string()
                        // naive parse for {"response":"..."}
                        val idx = text.indexOf("\"response\":\"")
                        if (idx >= 0) {
                            val start = text.indexOf('"', idx + 12)
                            val end = text.indexOf('"', start + 1)
                            val resp = if (start >= 0 && end > start) text.substring(start + 1, end) else text
                            onChunk(StreamResponse(resp, true))
                        } else onChunk(StreamResponse(text, true))
                        return
                    }
                    var buffer = StringBuilder()
                    while (!src.exhausted()) {
                        val line = src.readUtf8Line() ?: break
                        val trimmed = line.trim()
                        if (trimmed.isEmpty()) continue
                        if (trimmed.startsWith("data: ")) {
                            val jsonText = trimmed.substring(6).trim()
                            if (jsonText.isEmpty() || jsonText == "[DONE]") { onChunk(StreamResponse("", true)); return }
                            if (jsonText.contains("\"response\"")) {
                                val idx = jsonText.indexOf("\"response\":\"")
                                if (idx >= 0) {
                                    val s = jsonText.indexOf('"', idx + 12)
                                    val e = if (s >= 0) jsonText.indexOf('"', s + 1) else -1
                                    if (s >= 0 && e > s) onChunk(StreamResponse(jsonText.substring(s + 1, e), jsonText.contains("\"done\":true")))
                                }
                            } else {
                                onChunk(StreamResponse(jsonText, false))
                            }
                        } else {
                            onChunk(StreamResponse(trimmed, false))
                        }
                    }
                }
            }
        })
    }
}

